File: src/Blokemon.Product/DomainValues.cs
namespace Blokemon.Product;

public enum TextValueFailure
{
    Required,
}

public sealed record ProfileId
{
    private ProfileId(string value) => Value = value;

    public string Value { get; }

    public static DomainResult<ProfileId, TextValueFailure> Create(string? value) =>
        NonBlankText.Create(value, static valid => new ProfileId(valid));

    public override string ToString() => Value;
}

public sealed record DisplayName
{
    public const int MaximumLength = 32;

    private DisplayName(string value) => Value = value;

    public string Value { get; }

    public static DomainResult<DisplayName, DisplayNameCreationFailure> Create(string? value)
    {
        var trimmed = value?.Trim();
        if (string.IsNullOrEmpty(trimmed))
        {
            return DomainResult<DisplayName, DisplayNameCreationFailure>.Failure(
                DisplayNameCreationFailure.Required
            );
        }

        return trimmed.Length > MaximumLength
            ? DomainResult<DisplayName, DisplayNameCreationFailure>.Failure(
                DisplayNameCreationFailure.TooLong
            )
            : DomainResult<DisplayName, DisplayNameCreationFailure>.Success(
                new DisplayName(trimmed)
            );
    }

    public override string ToString() => Value;
}

public enum DisplayNameCreationFailure
{
    Required,
    TooLong,
}

public sealed record CommandId
{
    private CommandId(string value) => Value = value;

    public string Value { get; }

    public static DomainResult<CommandId, TextValueFailure> Create(string? value) =>
        NonBlankText.Create(value, static valid => new CommandId(valid));

    public override string ToString() => Value;
}

public sealed record PackReceiptId
{
    private PackReceiptId(string value) => Value = value;

    public string Value { get; }

    public static DomainResult<PackReceiptId, TextValueFailure> Create(string? value) =>
        NonBlankText.Create(value, static valid => new PackReceiptId(valid));

    public override string ToString() => Value;
}

public sealed record DeckId
{
    private DeckId(string value) => Value = value;

    public string Value { get; }

    public static DomainResult<DeckId, TextValueFailure> Create(string? value) =>
        NonBlankText.Create(value, static valid => new DeckId(valid));

    public override string ToString() => Value;
}

public sealed record StarterDeckId
{
    private StarterDeckId(string value) => Value = value;

    public string Value { get; }

    public static DomainResult<StarterDeckId, TextValueFailure> Create(string? value) =>
        NonBlankText.Create(value, static valid => new StarterDeckId(valid));

    public override string ToString() => Value;
}

public sealed record DeckName
{
    private DeckName(string value) => Value = value;

    public string Value { get; }

    public static DomainResult<DeckName, TextValueFailure> Create(string? value) =>
        NonBlankText.Create(value, static valid => new DeckName(valid));

    public override string ToString() => Value;
}

public sealed record CardId
{
    private CardId(string value) => Value = value;

    public string Value { get; }

    public static DomainResult<CardId, TextValueFailure> Create(string? value) =>
        NonBlankText.Create(value, static valid => new CardId(valid));

    internal static CardId FromAuthority(string value) => new(value);

    public override string ToString() => Value;
}

internal static class NonBlankText
{
    internal static DomainResult<TValue, TextValueFailure> Create<TValue>(
        string? value,
        Func<string, TValue> create
    )
        where TValue : notnull =>
        string.IsNullOrWhiteSpace(value)
            ? DomainResult<TValue, TextValueFailure>.Failure(TextValueFailure.Required)
            : DomainResult<TValue, TextValueFailure>.Success(create(value));
}

File: src/Blokemon.Product/DomainValues.fs
namespace Blokemon.Product

open System

/// Why a piece of required text was rejected.
type TextValueFailure =
    | Required = 0

/// Why a display name was rejected.
type DisplayNameCreationFailure =
    | Required = 0
    | TooLong = 1

/// Every identity below is the same shape: non-blank text, or a typed failure.
module internal NonBlankText =

    let create
        (value: string | null)
        (make: string -> 'TValue)
        : DomainResult<'TValue, TextValueFailure> =
        match value with
        | null -> DomainResult.Failed TextValueFailure.Required
        | text when String.IsNullOrWhiteSpace text -> DomainResult.Failed TextValueFailure.Required
        | text -> DomainResult.Succeeded(make text)

/// Identifies a local profile.
type ProfileId =
    private
        { value: string }

    member this.Value = this.value

    override this.ToString() = this.value

    static member op_Equality(left: ProfileId, right: ProfileId) = left.Equals(right)

    static member op_Inequality(left: ProfileId, right: ProfileId) = not (left.Equals(right))

    static member Create(value: string | null) =
        NonBlankText.create value (fun valid -> { value = valid })

/// The name a player chose for themselves.
type DisplayName =
    private
        { value: string }

    /// The longest display name the game accepts.
    static member MaximumLength = 32

    member this.Value = this.value

    override this.ToString() = this.value

    static member op_Equality(left: DisplayName, right: DisplayName) = left.Equals(right)

    static member op_Inequality(left: DisplayName, right: DisplayName) = not (left.Equals(right))

    static member Create
        (value: string | null)
        : DomainResult<DisplayName, DisplayNameCreationFailure> =
        let trimmed =
            match value with
            | null -> null
            | text -> text.Trim()

        match trimmed with
        | null -> DomainResult.Failed DisplayNameCreationFailure.Required
        | "" -> DomainResult.Failed DisplayNameCreationFailure.Required
        | text when text.Length > DisplayName.MaximumLength ->
            DomainResult.Failed DisplayNameCreationFailure.TooLong
        | text -> DomainResult.Succeeded { value = text }

/// Identifies the client request that caused a change.
type CommandId =
    private
        { value: string }

    member this.Value = this.value

    override this.ToString() = this.value

    static member op_Equality(left: CommandId, right: CommandId) = left.Equals(right)

    static member op_Inequality(left: CommandId, right: CommandId) = not (left.Equals(right))

    static member Create(value: string | null) =
        NonBlankText.create value (fun valid -> { value = valid })

/// Identifies one opened pack.
type PackReceiptId =
    private
        { value: string }

    member this.Value = this.value

    override this.ToString() = this.value

    static member op_Equality(left: PackReceiptId, right: PackReceiptId) = left.Equals(right)

    static member op_Inequality(left: PackReceiptId, right: PackReceiptId) =
        not (left.Equals(right))

    static member Create(value: string | null) =
        NonBlankText.create value (fun valid -> { value = valid })

/// Identifies a saved deck.
type DeckId =
    private
        { value: string }

    member this.Value = this.value

    override this.ToString() = this.value

    static member op_Equality(left: DeckId, right: DeckId) = left.Equals(right)

    static member op_Inequality(left: DeckId, right: DeckId) = not (left.Equals(right))

    static member Create(value: string | null) =
        NonBlankText.create value (fun valid -> { value = valid })

/// Identifies a starter deck in the catalogue.
type StarterDeckId =
    private
        { value: string }

    member this.Value = this.value

    override this.ToString() = this.value

    static member op_Equality(left: StarterDeckId, right: StarterDeckId) = left.Equals(right)

    static member op_Inequality(left: StarterDeckId, right: StarterDeckId) =
        not (left.Equals(right))

    static member Create(value: string | null) =
        NonBlankText.create value (fun valid -> { value = valid })

/// The name a player gave a saved deck.
type DeckName =
    private
        { value: string }

    member this.Value = this.value

    override this.ToString() = this.value

    static member op_Equality(left: DeckName, right: DeckName) = left.Equals(right)

    static member op_Inequality(left: DeckName, right: DeckName) = not (left.Equals(right))

    static member Create(value: string | null) =
        NonBlankText.create value (fun valid -> { value = valid })

/// Identifies a card in the mechanical authority.
type CardId =
    private
        { value: string }

    member this.Value = this.value

    override this.ToString() = this.value

    static member op_Equality(left: CardId, right: CardId) = left.Equals(right)

    static member op_Inequality(left: CardId, right: CardId) = not (left.Equals(right))

    static member Create(value: string | null) =
        NonBlankText.create value (fun valid -> { value = valid })

    static member internal FromAuthority(value: string) = { value = value }
